src/coreclr/jit/llvm.cpp

```diff
@@ -1654,14 +1490,153 @@ void startImportingNode(llvm::IRBuilder<>& builder)
+        GenTreeIntCon* offset = _compiler->gtNewIconNode(varDsc->GetStackOffset(), TYP_I_IMPL);
+        GenTreeLclVar* shadowStackVar = _compiler->gtNewLclvNode(_shadowStackLclNum, var_types::TYP_I_IMPL);
+        GenTree* lclAddress = _compiler->gtNewOperNode(genTreeOps::GT_ADD, var_types::TYP_I_IMPL, shadowStackVar, offset);
+
+        genTreeOps indirOper = GT_NONE;
+        GenTree* storedValue = nullptr;
+        switch (node->OperGet())
+        {
+        case GT_STORE_LCL_VAR:
+            indirOper = lclVar->TypeIs(TYP_STRUCT) ? GT_STORE_OBJ : GT_STOREIND;
+            storedValue = node->AsOp()->gtGetOp1();
+            break;
+        case GT_LCL_VAR:
+            indirOper = lclVar->TypeIs(TYP_STRUCT) ? GT_OBJ : GT_IND;
+            break;
+        }
+        node->ChangeOper(indirOper);
+        node->AsIndir()->SetAddr(lclAddress);
+        if (GenTree::OperIsStore(indirOper))
+        {
+            node->gtFlags |= GTF_IND_TGT_NOT_HEAP;
+            node->AsOp()->gtOp2 = storedValue;
+        }
+
+        CurrentRange().InsertBefore(node, offset);
+        CurrentRange().InsertAfter(offset, shadowStackVar);
+        CurrentRange().InsertAfter(shadowStackVar, lclAddress);
+    }
+}
+
+void Llvm::ConvertShadowStackLocals()
+{
+    _shadowStackLclNum = _compiler->lvaGrabTemp(true DEBUGARG("shadowstack"));
+    LclVarDsc* shadowStackVarDsc = _compiler->lvaGetDesc(_shadowStackLclNum);
+    shadowStackVarDsc->lvIsParam = 1;
+    shadowStackVarDsc->lvType = var_types::TYP_I_IMPL;
+
+    _compiler->eeGetMethodSig(_compiler->info.compMethodHnd, &_sigInfo);
+    _sigInfoIsValid = true;
+
+    for (BasicBlock* const block : _compiler->Blocks())
+    {
+        LIR::Range& currentRange = LIR::AsRange(block);
+        _currentRange = &LIR::AsRange(block);
+        for (GenTree* node : CurrentRange())
+        {
+            if (node->OperIs(GT_STORE_LCL_VAR, GT_LCL_VAR))
+            {
+                CreateShadowStackLocalAddress(node);
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------
+// Convert GT_STORE_LCL_VAR and GT_LCL_VAR to use the shadow stack when the local needs to be GC tracked
+//
+void Llvm::PlaceAndConvertShadowStackLocals()
+{
+    _shadowStackLocalsSize = 0;
+    if (_compiler->lvaCount == 0)
+    {
+        return;
+    }
+
+    std::vector<LclVarDsc*> locals;
+
+    unsigned lclNum;
+    LclVarDsc* varDsc;
+    for (lclNum = 0, varDsc = _compiler->lvaTable; lclNum < _compiler->lvaCount; lclNum++, varDsc++)
+    {
+        if (!canStoreLocalOnLlvmStack(varDsc))
+        {
+            locals.push_back(varDsc);
+        }
+    }
+
+    if (locals.size() == 0)
+    {
+        return;
+    }
+
+    if (_compiler->opts.OptimizationEnabled())
+    {
+        std::sort(locals.begin(), locals.end(), [](const LclVarDsc* lhs, const LclVarDsc* rhs) { return lhs->lvRefCntWtd() > rhs->lvRefCntWtd(); });
+    }
```

Comment:
We'll be sorting all variables... but don't arguments need to be at fixed offsets?